src/plugins/ipt/ipt.cpp

```diff
@@ -0,0 +1,469 @@
+    ef_data->plugin = plugin;
+    ef_data->rip = ((params->fault_va >> 12) << 12);
+    drakvuf_trap_t* exec_trap = (drakvuf_trap_t*)malloc(sizeof(drakvuf_trap_t));
+
+    exec_trap->type = MEMACCESS;
+    exec_trap->memaccess.gfn = p_info.paddr >> 12;
+    exec_trap->memaccess.type = PRE;
+    exec_trap->memaccess.access = VMI_MEMACCESS_X;
+    exec_trap->data = ef_data; // FIXME memleak
+    exec_trap->cb = execute_faulted_cb;
+    exec_trap->name = nullptr;
+
+    drakvuf_add_trap(drakvuf, exec_trap);
+    PRINT_DEBUG("[IPT] Trap X on GFN 0x%lx\n", p_info.paddr >> 12);
+
+    drakvuf_release_vmi(drakvuf);
+
+    plugin->destroy_trap(info->trap);
+
+    return VMI_EVENT_RESPONSE_NONE;
+}
+
+static event_response_t mm_access_fault_hook_cb(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    addr_t fault_va = drakvuf_get_function_argument(drakvuf, info, 2);
+    PRINT_DEBUG("[IPT] MmAccessFault(%d, %lx)\n", info->proc_data.pid, fault_va);
+
+    if (fault_va & (1ULL << 63))
+    {
+        PRINT_DEBUG("[IPT] Don't trap in kernel %d %lx\n", info->proc_data.pid, fault_va);
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+
+    auto plugin = get_trap_plugin<ipt>(info);
+    if (!plugin)
+        return VMI_EVENT_RESPONSE_NONE;
+
+    auto trap = plugin->register_trap<access_fault_result_t>(
+                    info,
+                    mm_access_fault_return_hook_cb,
+                    breakpoint_by_pid_searcher());
+    if (!trap)
+        return VMI_EVENT_RESPONSE_NONE;
+
+    auto params = get_trap_params<access_fault_result_t>(trap);
+    params->set_result_call_params(info, drakvuf_get_current_thread(drakvuf, info));
+    params->fault_va = fault_va;
+
+    return VMI_EVENT_RESPONSE_NONE;
+}
+
+event_response_t ipt_cr3_cb(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    uint32_t payloads[4] =
+    {
+        PTW_CURRENT_CR3, (uint32_t) info->regs->cr3,
+        PTW_CURRENT_TID, info->proc_data.tid
+    };
+
+    annotate_ipt(drakvuf, (ipt*)info->trap->data, info->vcpu, payloads, 2);
+    return VMI_EVENT_RESPONSE_NONE;
+}
+
+event_response_t ipt_catchall_cb(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    uint32_t payloads[2] =
+    {
+        PTW_EVENT_ID, (uint32_t) info->event_uid
+    };
+
+    annotate_ipt(drakvuf, (ipt*)info->trap->data, info->vcpu, payloads, 1);
+    return VMI_EVENT_RESPONSE_NONE;
+}
+
+ipt::ipt(drakvuf_t drakvuf, const ipt_config* c, output_format_t output)
+    : pluginex(drakvuf, output)
+{
+    for (unsigned int i = 0; i < this->num_vcpus; i++)
+    {
+        this->vcpu[i].fd = 0;
+    }
+
+    this->ipt_dir = c->ipt_dir;
+
+    if (!this->ipt_dir)
+    {
+        PRINT_DEBUG("[IPT] Target directory not provided, not activating IPT plugin\n");
+        return;
+    }
+
+    {
+        std::stringstream ss;
+        ss << this->ipt_dir;
+        ss << "/frames";
+
+        std::string frame_dir = ss.str();
+
+        int res = mkdir(frame_dir.c_str(), S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH);
```

Comment:
std::filesystem tho...
😢

Reply:
Turns out there are some problems with multi-distro support of `<filesystem>` :|
https://github.com/tklengyel/drakvuf/pull/1000/checks?check_run_id=1388206841

I would rather keep this `mkdir` here for better portability.